packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -37,70 +42,113 @@ function areDefinedAndEqual<T>(
-  const results = [result];
-  if (result.parentResult) {
-    results.unshift(result.parentResult);
-  }
-  if (result.childResults?.length) {
-    results.push(
-      ...result.childResults.flatMap((childResult) =>
-        flattenResultTree(childResult)
-      )
-    );
+function getFoldedResults(result: ResultWithFolding): ResultWithFolding[] {
+  if (!result.childResults) {
+    return [];
   }
-  return results;
+  return result.childResults.flatMap((childResult) => [
+    childResult,
+    ...getFoldedResults(childResult),
+  ]);
 }
 
-function foldResult(
-  source: ResultWithFolding,
+function addChildren(
+  parent: ResultWithFolding,
   results: ResultWithFolding[],
   fields: FoldingFields
 ): FoldedResult {
-  const sourceChildValue = getChildField(source, fields);
+  const sourceChildValue = getChildField(parent, fields);
   return {
-    ...source,
+    ...parent,
     children: sourceChildValue
       ? results
           .filter((result) => {
             const isSameResultAsSource =
-              getChildField(result, fields) === getChildField(source, fields);
+              getChildField(result, fields) === getChildField(parent, fields);
             const isChildOfSource =
               getParentField(result, fields) === sourceChildValue;
             return isChildOfSource && !isSameResultAsSource;
           })
-          .map((result) => foldResult(result, results, fields))
+          .map((result) => addChildren(result, results, fields))
       : [],
   };
 }
 
-function createCollection(
+function createCollectionFromResults(
+  results: ResultWithFolding[],
+  fields: FoldingFields,
+  moreResultsAvailable: boolean
+): Collection | null {
+  const rootResult = results.find((result) => {
+    const hasNoParent = getParentField(result, fields) === undefined;
+    const isParentOfItself = areDefinedAndEqual(
+      getParentField(result, fields),
+      getChildField(result, fields)
+    );
+    return hasNoParent || isParentOfItself;
+  });
+
+  if (!rootResult) {
+    return null;
+  }
+
+  return {
+    ...addChildren(rootResult, results, fields),
+    moreResultsAvailable,
+    isLoadingMoreResults: false,
+  };
+}
+
+function createCollectionFromResult(
   relevantResult: ResultWithFolding,
-  fields: FoldingFields
-): FoldedResult {
-  const flattenedResults = removeDuplicates(
-    flattenResultTree(relevantResult),
+  fields: FoldingFields,
+  numberOfFoldedResults: number
+): Collection {
+  const foldedResults = getFoldedResults(relevantResult);
+
+  const moreResultsAvailable = foldedResults.length > numberOfFoldedResults;
+
+  const foldedResultsWithoutBuffer = moreResultsAvailable
+    ? foldedResults.slice(
+        0,
+        foldedResults.length - collectionMoreResultsAvailableBuffer
+      )
+    : foldedResults;
+
+  const parentResults = [relevantResult, ...foldedResults]
+    .filter((result) => result.parentResult)
+    .map((result) => result.parentResult!);
```

Comment:
The index fetches additional results and adds them in the `parentResult` field of most (or all) results when they have `foldingparent` set. In my experimentation, this did not seem to count towards `filterFieldRange`/`numberOfFoldedResults`.